layers.go

```diff
@@ -918,18 +919,26 @@ func (r *layerStore) load(lockedForWriting bool) (bool, error) {
 				layer.Flags = make(map[string]interface{})
 			}
 			if layerHasIncompleteFlag(layer) {
-				logrus.Warnf("Found incomplete layer %#v, deleting it", layer.ID)
-				err := r.deleteInternal(layer.ID)
-				if err != nil {
-					// Don't return the error immediately, because deleteInternal does not saveLayers();
-					// Even if deleting one incomplete layer fails, call saveLayers() so that other possible successfully
-					// deleted incomplete layers have their metadata correctly removed.
-					incompleteDeletionErrors = multierror.Append(incompleteDeletionErrors,
-						fmt.Errorf("deleting layer %#v: %w", layer.ID, err))
-				}
+				// Important: Do not call r.deleteInternal() here. It modifies r.layers
+				// which causes unexpected side effects while iterating over r.layers here.
+				// The range loop has no idea that the underlying elements where shifted
+				// around.
+				layersToDeleteIDs = append(layersToDeleteIDs, layer.ID)
 				modifiedLocations |= layerLocation(layer)
```

Comment:
Even though it's ready only maybe this function call move too just for good measure?

Then conceptually of course this loop boils down to what in Rust would just be `r.layers.iter().filter(|layer| layer.is_incomplete())`

Reply:
I agree that moving this to the new loop would be cleaner. We would then need to collect `*Layer` pointers instead, but that should be fine; the loading code above has already ensured layer IDs are unique.